### PR TITLE
Move `id` and `json_api_type` methods from `Serializer` to `JsonApi`.

### PR DESCRIPTION
**Pull request by beauby**
Refactored `add_resource_relationship` along the way.

### CODE REVIEW

File: lib/active_model/serializer.rb [hunk 1]
```diff
@@ -146,18 +146,6 @@ def json_key
...
-
-    def json_api_type
-      if config.jsonapi_resource_type == :plural
-        object.class.model_name.plural
-      else
-        object.class.model_name.singular
-      end
-    end
-
     def attributes(options = {})
```
Comment (beauby) [inline lib/active_model/serializer.rb]: These two methods clearly did not belong there, as they are JSONAPI specific.

File: lib/active_model/serializer.rb [hunk 2]
```diff
@@ -146,18 +146,6 @@ def json_key
...
-    end
-
-    def json_api_type
-      if config.jsonapi_resource_type == :plural
-        object.class.model_name.plural
-      else
-        object.class.model_name.singular
-      end
-    end
-
```
Comment (beauby) [inline lib/active_model/serializer.rb]: These two methods clearly did not belong there, as they are JSONAPI specific.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 3]
```diff
@@ -44,19 +44,15 @@ def fragment_cache(cached_hash, non_cached_hash)
...
-          if serializer && serializer.object
-            resource[:relationships][name][:data] = { type: serializer.json_api_type, id: serializer.id.to_s }
-          end
+        def resource_identifier(serializer)
+          type = if ActiveModel::Serializer.config.jsonapi_resource_type == :plural
+                   serializer.object.class.model_name.plural
+                 else
+                   serializer.object.class.model_name.singular
+                 end
+          id = serializer.object.id.to_s
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: This might be easier to read as a case statement

```ruby
type = 
  case
  when  ActiveModel::Serializer.config.jsonapi_resource_type == :plural
    serializer.object.class.model_name.plural
  else
    serializer.object.class.model_name.singular
  end
```

maybe `resource_type` and `resource_id` as the local variable names?  

This method seems like there's a latent object in here that transforms a single resource into a serializable hash

Nice on the naming  http://jsonapi.org/format/#document-resource-identifier-objects :)

File: lib/active_model/serializer/adapter/json_api.rb [hunk 4]
```diff
@@ -100,15 +96,12 @@ def attributes_for_serializer(serializer, options)
...
 
           cache_check(serializer) do
             attributes = serializer.attributes(options)
+            attributes.delete(:id)
 
-            result = {
-              id: attributes.delete(:id).to_s,
-              type: attributes.delete(:json_api_type)
-            }
+            result = resource_identifier(serializer)
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I missed why you're throwing out `seriarlizer.attributes(options)[:id]`?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 5]
```diff
@@ -136,22 +129,24 @@ def check_assoc(assoc)
...
+                      serializer.map { |s| resource_identifier(s) }
+                    else
+                      if opts[:virtual_value]
+                        opts[:virtual_value]
+                      elsif serializer && serializer.object
+                        resource_identifier(serializer)
+                      else
+                        nil
+                      end
+                    end
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: yeah, definitely a latent object in here for serializer and serializer collections...  it would also make naming blocks easier e.g. `serializers.each { |serializer| }`

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: (That's OT, sorry... just thinking out loud)

File: lib/active_model/serializer/adapter/json_api.rb (hunk 4 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Because we do not want it to be part of the `attributes` property of the resource, but on the top-level alongside `type`.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 6]
```diff
@@ -100,15 +96,12 @@ def attributes_for_serializer(serializer, options)
 
         def resource_object_for(serializer, options)
           options[:fields] = @fieldset && @fieldset.fields_for(serializer)
-          options[:required_fields] = [:id, :json_api_type]
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: So, we don't need required fields because they're the `{ id: '', type: ''}` in the resource identifier object method?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 5 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I agree for naming blocks but here there are already two local variables named `serializer` (and in an other PR I'd actually like to use the outside one, so any suggestion on how to handle the naming situation?).

File: test/serializers/attributes_test.rb [hunk 7]
```diff
@@ -23,12 +23,6 @@ def test_attributes_with_fields_option
                      @profile_serializer.attributes(fields: [:name]))
       end
 
-      def test_required_fields
-        assert_equal({name: 'Name 1', description: 'Description 1'},
-                     @profile_serializer.attributes(fields: [:name, :description], required_fields: [:name]))
-
-      end
-
```
Comment (bf4) [inline test/serializers/attributes_test.rb]: This makes it look like you could set `:name` as a required_field.  Is that still possible?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 6 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Exactly.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 4 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: No, I get that.. but you're not using it, but it's there... that's what I mean by throwing it out... I'm sure there's a detail I'm missing in the refactor, but figured it'd be good to ask from ignorance :)

File: test/serializers/attributes_test.rb (hunk 7 above)
Comment (beauby) [inline test/serializers/attributes_test.rb]: What you imply is true, and the answer is no, it is not possible anymore. However, this option is not used anywhere inside of AMS, and I am quite unsure as to why one would use it outside considering its only use was to add attributes without declaring them as attributes. I'm convinced it was created as a hack for the sole purpose of handling the `json_api_type` and `id`.

File: lib/active_model/serializer/adapter/json_api.rb (hunk 3 above)
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Agreed for `resource_type` and `resource_id`.
Could you elaborate about this "latent object"?

File: lib/active_model/serializer/adapter/json_api.rb (hunk 4 above)
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: I mean, is the serializer.attributes(options)[:id] the same as serializer.object.id

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: No, good catch. This change actually breaks `id` overriding.

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: It is fixed now.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 8]
```diff
@@ -133,29 +131,29 @@ def check_assoc(assoc)
...
-
-            attrs[:relationships] ||= {}
-
-            if serializer.respond_to?(:each)
-              add_relationships(attrs, key, serializer)
+        def resource_relationship_value(serializer, options = {})
+          if serializer.respond_to?(:each)
+            serializer.map { |s| resource_identifier(s) }
+          else
+            if options[:virtual_value]
```
Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: After giving it a thought, why would anyone *ever* use a virtual value for a relationship with json_api? The only value this can take is `nil` or a Resource Identifier/array of Resource Identifiers, which means the virtual attribute has to be either of those, which means the use case is something like:
```ruby
PostSerializer < ActiveModel::Serializer
  attributes :id, :name
  belongs_to :author

  def author
    { type: 'authors', id: '3' }
  end
end
```

which in turn implies that one can't really include the corresponding resource.

Hence, I vote for removing the virtual values for relationships with json_api adapter.

File: lib/active_model/serializer.rb (hunk 2 above)
Comment (joaomdmoura) [inline lib/active_model/serializer.rb]: :+1: I'm happy we finally did it!

File: lib/active_model/serializer/adapter/json_api.rb [hunk 9]
```diff
@@ -44,21 +44,29 @@ def fragment_cache(cached_hash, non_cached_hash)
...
         end
 
-        def add_relationship(resource, name, serializer, val=nil)
-          resource[:relationships] ||= {}
-          resource[:relationships][name] = { data: val }
-
-          if serializer && serializer.object
-            resource[:relationships][name][:data] = { type: serializer.json_api_type, id: serializer.id.to_s }
+        def resource_identifier_id(serializer)
+          if serializer.respond_to?('id')
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: We are also adding possibility to override it then?

File: lib/active_model/serializer/adapter/json_api.rb [hunk 10]
```diff
@@ -44,21 +44,29 @@ def fragment_cache(cached_hash, non_cached_hash)
...
+            serializer.id.to_s
+          else
+            serializer.object.id.to_s
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: Nice touch converting it to string in order to follow json-api conventions

File: lib/active_model/serializer/adapter/json_api.rb [hunk 11]
```diff
@@ -88,28 +96,18 @@ def add_included(resource_name, serializers, parent = nil)
 
         def attributes_for_serializer(serializer, options)
           if serializer.respond_to?(:each)
-            result = []
-            serializer.each do |object|
-              result << resource_object_for(object, options)
-            end
+            serializer.map { |s| resource_object_for(s, options) }
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: :+1:

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: I'm working on the final bit to improve handling of the "included"
relationships. After that, it should be pretty clean/understandable/easy to
work with.

Le mardi 1 septembre 2015, João Moura <notifications@github.com> a écrit :

> In lib/active_model/serializer/adapter/json_api.rb
> <https://github.com/rails-api/active_model_serializers/pull/1103#discussion_r38415770>
> :
>
> > @@ -88,28 +96,18 @@ def add_included(resource_name, serializers, parent = nil)
> >
> >          def attributes_for_serializer(serializer, options)
> >            if serializer.respond_to?(:each)
> > -            result = []
> > -            serializer.each do |object|
> > -              result << resource_object_for(object, options)
> > -            end
> > +            serializer.map { |s| resource_object_for(s, options) }
>
> [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails-api/active_model_serializers/pull/1103/files#r38415770>
> .
>


-- 
Lucas Hosseini
lucas.hosseini@gmail.com

File: lib/active_model/serializer/adapter/json_api.rb (hunk 8 above)
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: Maybe we could do something similar to JsonAdapter, that if found a instance of an model object it will query that object and use it as the relationship. I know that it might be an edge case, but would be nice to cover, taking into account the level o abstraction we are working with when trying to cover everyone using AMS :grin:

File: test/serializers/attributes_test.rb (hunk 7 above)
Comment (joaomdmoura) [inline test/serializers/attributes_test.rb]: Despite of felling okay removing this, it just remembers me how badly we need the filter method hahah because handle with different set `fields` using the same serializer is a need.
Btw as I commented [here](https://github.com/rails-api/active_model_serializers/pull/1103/files#r38414212), it seems we are adding some new behavior, would nice to have it tested :smile:

File: lib/active_model/serializer/adapter/json_api.rb (hunk 9 above)
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: I also see you some symbols on the other places and here a string, I'd go with a symbol as well

File: lib/active_model/serializer/adapter/json_api.rb [hunk 12]
```diff
@@ -133,29 +131,29 @@ def check_assoc(assoc)
...
-            serializer = association.serializer
-            opts = association.options
-
-            attrs[:relationships] ||= {}
-
-            if serializer.respond_to?(:each)
-              add_relationships(attrs, key, serializer)
+        def resource_relationship_value(serializer, options = {})
+          if serializer.respond_to?(:each)
+            serializer.map { |s| resource_identifier(s) }
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: I'm not sure if we have a test covering this to check if we are following json-api conventions, but would be nice to cover this one:
>...
> Resource linkage MUST be represented as one of the following:
>...
> - an empty array ([]) for empty to-many relationships.

Comment (beauby) [inline lib/active_model/serializer/adapter/json_api.rb]: Agreed. Will issue a PR for the corresponding test in a bit.

File: test/serializers/attributes_test.rb (hunk 7 above)
Comment (bf4) [inline test/serializers/attributes_test.rb]: I'd just prefer to keep breaking changes out of refactoring pr...

Comment (beauby) [inline test/serializers/attributes_test.rb]: As explained, this is not a breaking change, as this feature was not
actually a frature but merely a hack that happened to be exposed.

Le mercredi 2 septembre 2015, Benjamin Fleischer <notifications@github.com>
a écrit :

> In test/serializers/attributes_test.rb
> <https://github.com/rails-api/active_model_serializers/pull/1103#discussion_r38569520>
> :
>
> > @@ -23,12 +23,6 @@ def test_attributes_with_fields_option
> >                       @profile_serializer.attributes(fields: [:name]))
> >        end
> >
> > -      def test_required_fields
> > -        assert_equal({name: 'Name 1', description: 'Description 1'},
> > -                     @profile_serializer.attributes(fields: [:name, :description], required_fields: [:name]))
> > -
> > -      end
> > -
>
> I'd just prefer to keep breaking changes out of refactoring pr...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails-api/active_model_serializers/pull/1103/files#r38569520>
> .
>


-- 
Lucas Hosseini
lucas.hosseini@gmail.com

**Comment by beauby** [comment]
Rebased.

File: lib/active_model/serializer/adapter/json_api.rb [hunk 13]
```diff
@@ -9,7 +9,13 @@ def initialize(serializer, options = {})
           super
           @hash = { data: [] }
 
-          if fields = options.delete(:fields)
+          @options[:include] ||= []
+          if @options[:include].is_a?(String)
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: Can you explain more about this. It's not related to `type` and `id`, and I also don't understood the `String` verification

**Comment by joaomdmoura** [comment]
How do you feel about split the `add_resource_relationship` into a different PR? Both seems pretty solid, just want to make it easier to review :smile:

File: lib/active_model/serializer/adapter/json_api.rb [hunk 14]
```diff
@@ -31,142 +37,128 @@ def serializable_hash(options = nil)
...
+            @hash[:data] = primary_data
+            @hash[:data][:relationships] = relationships if relationships.any?
+            @hash[:included] = included if included.any?
           end
           @hash
         end
 
         def fragment_cache(cached_hash, non_cached_hash)
           root = false if @options.include?(:include)
-          JsonApi::FragmentCache.new().fragment_cache(root, cached_hash, non_cached_hash)
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: :+1:

File: lib/active_model/serializer/adapter/json_api.rb [hunk 15]
```diff
@@ -31,142 +37,128 @@ def serializable_hash(options = nil)
...
 
-        def add_included(resource_name, serializers, parent = nil)
-          unless serializers.respond_to?(:each)
-            return unless serializers.object
-            serializers = Array(serializers)
-          end
-          resource_path = [parent, resource_name].compact.join('.')
-          if include_assoc?(resource_path)
-            @hash[:included] ||= []
+        def resource_identifier_for(serializer)
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: :+1:

File: lib/active_model/serializer/adapter/json_api.rb [hunk 16]
```diff
@@ -31,142 +37,128 @@ def serializable_hash(options = nil)
...
 
-          serializers.each do |serializer|
-            serializer.associations.each do |association|
-              serializer = association.serializer
-
-              add_included(association.key, serializer, resource_path) if serializer
-            end if include_nested_assoc? resource_path
+        def primary_data_for(serializer, options)
+          if serializer.respond_to?(:each)
+            serializer.map { |s| resource_object_for(s, options) }
```
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: :+1:

File: lib/active_model/serializer/adapter/json_api.rb (hunk 13 above)
Comment (joaomdmoura) [inline lib/active_model/serializer/adapter/json_api.rb]: nvm, understood :grin:

**Comment by joaomdmoura** [comment]
LGTM, I'm merging it

**Comment by joaomdmoura** [comment]
Oh, could you please @beauby open a PR with a new 'how to' article explaning how to override `id` and `type`?

**Comment by beauby** [comment]
@joaomdmoura With this PR, `type` cannot be overridden (but a small change as in #1122 allows for that as well). Type can be overridden as one would expect it, by defining an `id` method on the serializer class. Maybe a little note in the existing doc would make more sense than a whole article.

**Comment by joaomdmoura** [comment]
The article size don't bother me and indeed it can even be a note, just want to make it easier to be found :smile:  but I'll check it on #1122
